devicetypes/smartthings/aeotec-wallmote.src/aeotec-wallmote.groovy

```diff
@@ -0,0 +1,171 @@
+            secure(zwave.wakeUpV2.wakeUpNoMoreInformation())
+    ])
+}
+
+def updated() {
+    if (!childDevices) {
+        createChildDevices()
+    }
+    else if (device.label != state.oldLabel) {
+        childDevices.each {
+            def segs = it.deviceNetworkId.split(":")
+            def newLabel = "${device.displayName} button ${segs[-1]}"
+            it.setLabel(newLabel)
+        }
+        state.oldLabel = device.label
+    }
+}
+
+def configure() {
+
+}
+
+def parse(String description) {
+    def results = []
+    if (description.startsWith("Err")) {
+        results = createEvent(descriptionText:description, displayed:true)
+    } else {
+        def cmd = zwave.parse(description)
+        if(cmd) results += zwaveEvent(cmd)
+        if(!results) results = [ descriptionText: cmd, displayed: false ]
+    }
+    return results
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.centralscenev1.CentralSceneNotification cmd) {
+    def button = cmd.sceneNumber
+    def value
+    // 0 = pushed, 1 = released, 2 = held down
+    if (cmd.keyAttributes != 2) {
+        value = cmd.keyAttributes == 1 ? "held" : "pushed"
+    } else {
+        // we can't do anything with the held down event yet
+        return []
+    }
+    def child = getChildDevice(button)
+    child?.sendEvent(name: "button", value: value, data: [buttonNumber: button], descriptionText: "$child.displayName was $value", isStateChange: true)
+    createEvent(name: "button", value: value, data: [buttonNumber: button], descriptionText: "$device.displayName button $button was $value", isStateChange: true)
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.securityv1.SecurityMessageEncapsulation cmd) {
+    def encapsulatedCommand = cmd.encapsulatedCommand()
+    if (encapsulatedCommand) {
+        return zwaveEvent(encapsulatedCommand)
+    } else {
+        log.warn "Unable to extract encapsulated cmd from $cmd"
+        createEvent(descriptionText: cmd.toString())
+    }
+}
+
+def zwaveEvent(physicalgraph.zwave.Command cmd) {
+    def linkText = device.label ?: device.name
+    [linkText: linkText, descriptionText: "$linkText: $cmd", displayed: false]
+}
+
+
+def zwaveEvent(physicalgraph.zwave.commands.wakeupv2.WakeUpNotification cmd) {
+    def results = []
+    results += createEvent(descriptionText: "$device.displayName woke up", isStateChange: false)
+    results += response([
+            secure(zwave.batteryV1.batteryGet()),
+            "delay 2000",
+            secure(zwave.wakeUpV2.wakeUpNoMoreInformation())
+    ])
+    results
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.batteryv1.BatteryReport cmd) {
+    def map = [ name: "battery", unit: "%", isStateChange: true ]
+    if (cmd.batteryLevel == 0xFF) {
+        map.value = 1
+        map.descriptionText = "$device.displayName battery is low!"
+    } else {
+        map.value = cmd.batteryLevel
+    }
+    createEvent(map)
+}
+
+def createChildDevices() {
+    state.oldLabel = device.label
+    def child
```

Comment:
Can move the `def` to inside the `for`

Reply:
doesn't that re-allocate on every loop?

Reply:
It's only used within the loop so no need to def it outside the scope of the loop

Reply:
I got that, but I'm trying to just allocate once instead of allocating `numberOfButtons` times.

Reply:
I don't know the internal details of Groovy but any overhead caused by this is going to be minuscule. Better to define the variables where they are needed.